src/core/experimental/memory/memcpy.d

```diff
@@ -0,0 +1,401 @@
+/**
+ * Pure D replacement of the C Standard Library memcpy().
+ * There is an idiomatic-D interface, memcpy(), which is split into 3 overloads.
+ * One taking static types, one dynamic arrays and one static arrays.
+ * Also, there is available a C-like interface, the Dmemcpy() (which is named Dmemcpy
+ * for disambiguation with the C memcpy() a _similar_ interface) that
+ * is the classic (void*, void*, size_t) interface.
+ * N.B.: Both the memcpy() here and Dmemcpy() return nothing, contrary to the C Standard
+ * Library version.
+ * Source: $(DRUNTIMESRC core/experimental/memory/memcpy.d)
+ */
+module core.experimental.memory.memcpy;
+
+import core.internal.traits : isArray;
+
+/**
+ * Handle Static Types
+ * N.B.: No need for more sophisticated code for static types. The compiler
+ * knows better how to handle them in every target case.
+ *
+ * Params:
+ *  dst = Reference to memory destination to copy bytes to.
+ *  src = Reference to memory source to copy bytes from.
+ */
+pragma(inline, true)
+void memcpy(T)(ref T dst, ref const T src)
+if (!isArray!T)
+{
+    dst = src;
+}
+
+/**
+ * Handle Dynamic Types
+ *
+ * Params:
+ *  dst = Reference to destination dynamic array to copy bytes to.
+ *  src = Reference to source dynamic array to copy bytes from.
+ */
+void memcpy(T)(ref T[] dst, ref const T[] src)
+{
+    assert(dst.length == src.length);
+    void* d = cast(void*) dst.ptr;
+    const(void)* s = cast(const(void)*) src.ptr;
+    size_t n = dst.length * typeof(dst[0]).sizeof;
+    // Assume that there is no overlap.
+    pragma(inline, true);
+    Dmemcpy(d, s, n);
+}
+
+/**
+ * Handle Dynamic Types
+ *
+ * Params:
+ *  len = Length of the static arrays.
+ *  dst = Reference to destination static array to copy bytes to.
+ *  src = Reference to source static array to copy bytes from.
+ */
+void memcpy(T, size_t len)(ref T[len] dst, ref const T[len] src)
+{
+    T[] d = dst[0 .. $];
+    const T[] s = src[0 .. $];
```

Comment:
Why write these on separate lines?

Reply:
`d` and `s`? How else could they be?

Reply:
`memcpy(dst[0 .. $], src[0 .. $]);`

I only say because I stopped and re-read these lines carefully looking for the gotcha I assumed must exist to justify this long-written version. To me, writing it out this way makes it look like there's unexpected detail that should be distinctly stated... maybe that's just me though. I was looking for the substance in the noise, and there wasn't any.

Reply:
I agree. And yes, there's no gotcha.

Reply:
I agree that writing it in one line expresses the intent clearer.